File: src/cleansia_android/core/src/main/java/cz/cleansia/core/ui/components/CleansiaDialog.kt
package cz.cleansia.core.ui.components

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.core.MutableTransitionState
import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.spring
import androidx.compose.animation.core.tween
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.scaleIn
import androidx.compose.animation.scaleOut
import androidx.compose.animation.slideInVertically
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.TransformOrigin
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties

/**
 * Cleansia-styled dialog. Replaces stock Material3 [androidx.compose.material3.AlertDialog]
 * across the app so confirm flows share one shape, padding, and typography.
 *
 * Two flavors via [destructive]: standard (primary-tinted confirm) and
 * destructive (error-tinted confirm + matching icon halo for delete/cancel).
 *
 * Optional [icon] renders a circular halo above the title — use for high-stakes
 * confirms (delete account, cancel subscription) where a glyph helps the user
 * orient. Omit for routine actions.
 *
 * For dialogs with custom body content (a TextField, a list of cleaners),
 * pass a [content] slot instead of [message]. When both are provided, [content]
 * renders below [message]. The Surface's [widthIn] keeps the dialog from
 * stretching edge-to-edge on tablets.
 */
@Composable
fun CleansiaDialog(
    onDismiss: () -> Unit,
    title: String,
    confirmLabel: String,
    onConfirm: () -> Unit,
    modifier: Modifier = Modifier,
    message: String? = null,
    dismissLabel: String? = null,
    icon: ImageVector? = null,
    destructive: Boolean = false,
    confirmEnabled: Boolean = true,
    content: (@Composable () -> Unit)? = null,
) {
    Dialog(
        onDismissRequest = onDismiss,
        properties = DialogProperties(usePlatformDefaultWidth = false),
    ) {
        // Entrance choreography: spring-driven scale-up from 0.85 around the
        // bottom anchor + a small upward slide + fade. The transition state
        // starts collapsed and flips to visible on the first composition so
        // the springy "pop" plays even though Dialog itself fades the window
        // in/out at the platform layer. Exit is a fast fade — Dialog tears
        // down the window the moment onDismiss flips state, so a slow exit
        // animation would never finish anyway.
        val visibleState = remember { MutableTransitionState(false) }
            .apply { targetState = true }

        AnimatedVisibility(
            visibleState = visibleState,
            enter = scaleIn(
                animationSpec = spring(
                    dampingRatio = Spring.DampingRatioMediumBouncy,
                    stiffness = Spring.StiffnessMediumLow,
                ),
                initialScale = 0.85f,
                transformOrigin = TransformOrigin(0.5f, 0.6f),
            ) + fadeIn(animationSpec = tween(durationMillis = 220)) +
                slideInVertically(
                    animationSpec = spring(
                        dampingRatio = Spring.DampingRatioMediumBouncy,
                        stiffness = Spring.StiffnessMediumLow,
                    ),
                    initialOffsetY = { it / 8 },
                ),
            exit = fadeOut(animationSpec = tween(durationMillis = 120)) +
                scaleOut(
                    animationSpec = tween(durationMillis = 120),
                    targetScale = 0.95f,
                ),
        ) {
        Surface(
            modifier = modifier
                .padding(horizontal = 24.dp)
                .widthIn(max = 420.dp)
                .fillMaxWidth(),
            shape = RoundedCornerShape(24.dp),
            color = MaterialTheme.colorScheme.surface,
            tonalElevation = 6.dp,
            shadowElevation = 12.dp,
        ) {
            Column(
                modifier = Modifier.padding(horizontal = 24.dp, vertical = 24.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
            ) {
                if (icon != null) {
                    val haloColor = if (destructive) {
                        MaterialTheme.colorScheme.errorContainer
                    } else {
                        MaterialTheme.colorScheme.primaryContainer
                    }
                    val tint = if (destructive) {
                        MaterialTheme.colorScheme.onErrorContainer
                    } else {
                        MaterialTheme.colorScheme.onPrimaryContainer
                    }
                    Box(
                        modifier = Modifier
                            .size(56.dp)
                            .background(haloColor, CircleShape),
                        contentAlignment = Alignment.Center,
                    ) {
                        Icon(
                            imageVector = icon,
                            contentDescription = null,
                            tint = tint,
                            modifier = Modifier.size(28.dp),
                        )
                    }
                    Spacer(Modifier.height(16.dp))
                }

                Text(
                    text = title,
                    style = MaterialTheme.typography.titleLarge,
                    color = MaterialTheme.colorScheme.onSurface,
                )

                if (message != null) {
                    Spacer(Modifier.height(8.dp))
                    Text(
                        text = message,
                        style = MaterialTheme.typography.bodyMedium,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }

                if (content != null) {
                    Spacer(Modifier.height(16.dp))
                    Box(modifier = Modifier.fillMaxWidth()) { content() }
                }

                Spacer(Modifier.height(24.dp))
                // Equal-weight filled buttons à la Wolt / Bolt confirms — both
                // sides read as tappable targets, the destructive side gets a
                // saturated fill so it never feels like the safe default.
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.spacedBy(12.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    if (dismissLabel != null) {
                        DialogActionButton(
                            label = dismissLabel,
                            onClick = onDismiss,
                            containerColor = MaterialTheme.colorScheme.surfaceVariant,
                            contentColor = MaterialTheme.colorScheme.onSurface,
                            modifier = Modifier.weight(1f),
                        )
                    }
                    val confirmContainer = if (destructive) {
                        MaterialTheme.colorScheme.error
                    } else {
                        MaterialTheme.colorScheme.primary
                    }
                    val confirmContent = if (destructive) {
                        MaterialTheme.colorScheme.onError
                    } else {
                        MaterialTheme.colorScheme.onPrimary
                    }
                    DialogActionButton(
                        label = confirmLabel,
                        onClick = onConfirm,
                        containerColor = confirmContainer,
                        contentColor = confirmContent,
                        enabled = confirmEnabled,
                        modifier = Modifier.weight(1f),
                    )
                }
            }
        }
        }
    }
}

@Composable
private fun DialogActionButton(
    label: String,
    onClick: () -> Unit,
    containerColor: androidx.compose.ui.graphics.Color,
    contentColor: androidx.compose.ui.graphics.Color,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
) {
    Button(
        onClick = onClick,
        enabled = enabled,
        modifier = modifier.defaultMinSize(minHeight = 48.dp),
        shape = RoundedCornerShape(14.dp),
        contentPadding = PaddingValues(horizontal = 16.dp, vertical = 10.dp),
        colors = ButtonDefaults.buttonColors(
            containerColor = containerColor,
            contentColor = contentColor,
        ),
    ) {
        // Material `Button` lays out its slot as a Row(Start) — wrap in a Box
        // so the label centers within the weighted button regardless of width.
        Box(
            modifier = Modifier.fillMaxWidth(),
            contentAlignment = Alignment.Center,
        ) {
            Text(
                text = label,
                style = MaterialTheme.typography.labelLarge,
            )
        }
    }
}

File: src/cleansia_ios/CleansiaCore/Sources/CleansiaCore/Components/CleansiaDialog.swift
import SwiftUI

public struct CleansiaDialog<Content: View>: View {
    private let title: String
    private let message: String?
    private let confirmLabel: String
    private let dismissLabel: String?
    private let icon: String?
    private let destructive: Bool
    private let confirmEnabled: Bool
    private let onConfirm: () -> Void
    private let onDismiss: () -> Void
    private let content: Content?

    @State private var presented = false

    public init(
        title: String,
        confirmLabel: String,
        onConfirm: @escaping () -> Void,
        onDismiss: @escaping () -> Void,
        message: String? = nil,
        dismissLabel: String? = nil,
        icon: String? = nil,
        destructive: Bool = false,
        confirmEnabled: Bool = true,
        @ViewBuilder content: () -> Content
    ) {
        self.title = title
        self.confirmLabel = confirmLabel
        self.onConfirm = onConfirm
        self.onDismiss = onDismiss
        self.message = message
        self.dismissLabel = dismissLabel
        self.icon = icon
        self.destructive = destructive
        self.confirmEnabled = confirmEnabled
        self.content = content()
    }

    public var body: some View {
        ZStack {
            Color.black.opacity(0.4)
                .ignoresSafeArea()
                .opacity(presented ? 1 : 0)
                .onTapGesture(perform: onDismiss)

            VStack(spacing: 0) {
                if let icon {
                    ZStack {
                        Circle()
                            .fill(destructive ? CleansiaColors.errorContainer : CleansiaColors.primaryContainer)
                            .frame(width: 56, height: 56)
                        Image(systemName: icon)
                            .font(.system(size: 28))
                            .foregroundColor(destructive ? CleansiaColors.onErrorContainer : CleansiaColors
                                .onPrimaryContainer)
                    }
                    .padding(.bottom, Spacing.m)
                }

                Text(title)
                    .font(CleansiaTypography.titleLarge)
                    .foregroundColor(CleansiaColors.onSurface)
                    .multilineTextAlignment(.center)

                if let message {
                    Text(message)
                        .font(CleansiaTypography.bodyMedium)
                        .foregroundColor(CleansiaColors.onSurfaceVariant)
                        .multilineTextAlignment(.center)
                        .padding(.top, Spacing.xs)
                }

                if let content {
                    content.padding(.top, Spacing.m)
                }

                HStack(spacing: Spacing.s) {
                    if let dismissLabel {
                        DialogButton(
                            label: dismissLabel,
                            background: CleansiaColors.surfaceVariant,
                            foreground: CleansiaColors.onSurface,
                            action: onDismiss
                        )
                    }
                    DialogButton(
                        label: confirmLabel,
                        background: destructive ? CleansiaColors.error : CleansiaColors.primary,
                        foreground: destructive ? CleansiaColors.onError : CleansiaColors.onPrimary,
                        enabled: confirmEnabled,
                        action: onConfirm
                    )
                }
                .padding(.top, Spacing.l)
            }
            .padding(Spacing.l)
            .frame(maxWidth: 420)
            .background(CleansiaColors.surface)
            .clipShape(RoundedRectangle(cornerRadius: CornerRadius.large))
            .shadow(radius: 24)
            .padding(.horizontal, Spacing.l)
            .scaleEffect(presented ? 1 : 0.85)
            .opacity(presented ? 1 : 0)
        }
        .onAppear {
            withAnimation(.spring(response: 0.4, dampingFraction: 0.62)) { presented = true }
        }
    }
}

public extension CleansiaDialog where Content == EmptyView {
    init(
        title: String,
        confirmLabel: String,
        onConfirm: @escaping () -> Void,
        onDismiss: @escaping () -> Void,
        message: String? = nil,
        dismissLabel: String? = nil,
        icon: String? = nil,
        destructive: Bool = false,
        confirmEnabled: Bool = true
    ) {
        self.init(
            title: title,
            confirmLabel: confirmLabel,
            onConfirm: onConfirm,
            onDismiss: onDismiss,
            message: message,
            dismissLabel: dismissLabel,
            icon: icon,
            destructive: destructive,
            confirmEnabled: confirmEnabled,
            content: { EmptyView() }
        )
    }
}

private struct DialogButton: View {
    let label: String
    let background: Color
    let foreground: Color
    var enabled = true
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            Text(label)
                .font(CleansiaTypography.labelLarge)
                .foregroundColor(foreground)
                .frame(maxWidth: .infinity, minHeight: 48)
                .background(background.opacity(enabled ? 1 : 0.5))
                .clipShape(RoundedRectangle(cornerRadius: 14))
        }
        .buttonStyle(.plain)
        .disabled(!enabled)
    }
}

#if DEBUG
    struct CleansiaDialog_Previews: PreviewProvider {
        static var previews: some View {
            CleansiaDialog(
                title: "Delete account?",
                confirmLabel: "Delete",
                onConfirm: {},
                onDismiss: {},
                message: "This permanently removes your account and data.",
                dismissLabel: "Cancel",
                icon: "trash",
                destructive: true
            )
        }
    }
#endif
